File: 1-working-with-list/p20_remove_at.ml

let remove_at n l =
  let rec aux acc curr = function
  | [] -> List.rev acc
  | h :: t ->
    if curr = 0 then
      (List.rev acc) @ t
    else
      aux (h::acc) (curr-1) t
  in aux [] n l;;

remove_at 1 ["a";"b";"c";"d"];;
remove_at 2 ["a";"b";"c";"d"];;
remove_at 0 ["a";"b";"c";"d"];;
remove_at 10 ["a";"b";"c";"d"];;
remove_at (-2) ["a";"b";"c";"d"];;

(* not tail recursive *)
let rec remove_at n = function
  | [] -> []
  | h :: t ->
    if n = 0 then
      t
    else
      h :: remove_at (n-1) t;;

remove_at 1 ["a";"b";"c";"d"];;
remove_at 2 ["a";"b";"c";"d"];;
remove_at 0 ["a";"b";"c";"d"];;
remove_at 10 ["a";"b";"c";"d"];;
remove_at (-2) ["a";"b";"c";"d"];;

File: 1-working-with-list/p20_remove_at.scala

object Main {

  def main(args: Array[String]) {
    val l = List("a", "b", "c", "d")
    println(removeAt(1, l))
    println(removeAt(0, l))
    println(removeAt(3, l))
    println(removeAt(6, l))
    println(removeAt2(1, l))
    println(removeAt2(0, l))
    println(removeAt2(3, l))
    println(removeAt2(6, l))
  }

  def removeAt[T](index: Int, list: List[T]): List[T] = {
    def aux(l: List[T], c: Int, acc: List[T]): List[T] = {
      l match {
        case Nil => acc.reverse
        case hd :: tail =>
          if(c == 0)
            acc.reverse ::: tail
          else
            aux(tail, c-1, hd :: acc)
      }
    }
    aux(list, index, List())
  }
  def removeAt2[T](index: Int, list: List[T]): List[T] = {
    list match {
      case Nil => list
      case hd :: tail =>
        if(index == 0)
          tail
        else
          hd :: removeAt2(index-1, tail)
    }
  }
}
